torchao/quantization/quantize_/int4/int4_groupwise_preshuffle_tensor.py

```diff
@@ -77,36 +99,53 @@ def __new__(
 
     def __init__(
         self,
-        packed_weight,
-        group_scale,
-        group_zero,
-        group_size,
-        shape_multiplier,
-        shape,
+        packed_weight: torch.Tensor,
+        group_scale: torch.Tensor,
+        group_zero: Optional[torch.Tensor],
+        row_scale: Optional[torch.Tensor],
+        group_size: int,
+        shape_multiplier: List[int],
+        shape: List[int],
     ):
         self.packed_weight = packed_weight
         self.group_scale = group_scale
         self.group_zero = group_zero
+        self.row_scale = row_scale
         self.shape_multiplier = shape_multiplier
         self.group_size = group_size
 
     def __tensor_flatten__(self):
-        return self.tensor_data_attrs, [
-            getattr(self, attr) for attr in self.tensor_attributes
-        ]
+        if getattr(self, self.optional_tensor_data_attr1) is None:
+            assert getattr(self, self.optional_tensor_data_attr2) is not None
+            return self.tensor_data_attrs + [self.optional_tensor_data_attr2], [
+                getattr(self, attr) for attr in self.tensor_attributes
+            ]
+        else:
+            assert getattr(self, self.optional_tensor_data_attr1) is not None
+            return self.tensor_data_attrs + [self.optional_tensor_data_attr1], [
+                getattr(self, attr) for attr in self.tensor_attributes
+            ]
 
     @classmethod
     def __tensor_unflatten__(
         cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride
     ):
+        tensors = [tensor_data_dict[name] for name in cls.tensor_data_attrs]
+        tensors.append(tensor_data_dict.get(cls.optional_tensor_data_attr1, None))
+        tensors.append(tensor_data_dict.get(cls.optional_tensor_data_attr2, None))
         return cls(
-            *[tensor_data_dict[name] for name in cls.tensor_data_attrs],
+            *tensors,
             *tensor_attributes,
         )
 
     def _apply_fn_to_data(self, fn):
+        tensors = [fn(getattr(self, name)) for name in self.tensor_data_attrs]
+        t1 = getattr(self, self.optional_tensor_data_attr1)
```

Comment:
seems like they should also go in a list?

Reply:
it can be a list of exactly two elements, that if one exists the other will not exist. it's a bit weird anyways, I'm open to just hardcode these as well, which one do you feel is better?